### ISSUE TITLE
Regression: FIDO2 key ignored for unlocking LUKS root using systemd-gpt-auto-generator

### ISSUE THREAD

**Issue by kjkent** [dracut-ng/dracut-ng#506]
**Describe the bug**
When using [systemd-gpt-auto-generator ](https://www.freedesktop.org/software/systemd/man/latest/systemd-gpt-auto-generator.html) to implement the [Discoverable Partitions Specification](https://uapi-group.org/specifications/specs/discoverable_partitions_specification/), Dracut does not consider a `systemd-cryptenroll`d' FIDO2 token to unlock the root volume.  This setup previously worked, but since spending a few months running Booster, my previous Dracut config and any permutations do not function as expected. One change is that I've swapped to building UKIs with Dracut.

Related issue on systemd:  https://github.com/systemd/systemd/issues/32586. I have noticed that user utilises their TPM for measured boot, whereas I'm yet to implement that. Therefore, there's no mention of the TPM within my config, though I get the same log output regarding the TPM and JSON tokens. I'm unsure if the poster of that issue (@capito27) is building a UKI.

Grepping the initrd logs/journalctl/dmesg for FIDO2/fido/YubiKey/ returns next to nothing; the only mentions are their presence in the kernel args and the udev handling of the YubiKey.

**Config**
- `/etc/dracut.conf.d/dracut.conf` _I've also tried toggling the `hostonly` opts_
```
hostonly='yes'
# includes booted kernel args that conflict with defined
hostonly_cmdline='no'
machine_id='no'
uefi='yes'
uefi_splash_image='/etc/dracut.conf.d/splash.bmp'
show_modules='yes'
enhanced_cpio='yes'
parallel='yes'

# misc
omit_dracutmodules+=' brltty '
kernel_cmdline+=' root=gpt-auto rd.auto=1 rd.systemd.gpt_auto=1 systemd.gpt_auto=1 '
kernel_cmdline+=' zswap.enabled=0 rd.luks.allow-discards '
kernel_cmdline+=' module_blacklist=sp5100_tco,wacom nowatchdog '
```

- `/etc/dracut.conf.d/fido2.conf` _I've also tried every variation I can think of, starting from simply adding the `fido2` dracut module and `rd.luks.options=fido2-device=auto` (which was my previous working config)._
```
force_add_dracutmodules+=' fido2 '
add_dracutmodules+=' fido2 '
force_drivers+=' usbhid hid_sensor_hub '
omit_dracutmodules+=' systemd.pcrphase tpm2-tss '
kernel_cmdline+=' rd.luks.options=fido2-device=auto rd.driver.pre=usbhid,hid_sensor_hub '
```

- `/etc/dracut.conf.d/debug.conf`
```
kernel_cmdline+=' rd.debug rd.shell systemd.log_level=debug systemd.debug_shell=1 '
```

Alongside the above config, systemd-boot is used as a bootloader, with sbctl used to sign the `.efi` files.

**Distribution used**
Arch Linux

**Dracut version**
102, 103

**Init system**
systemd (v256)

**To Reproduce**
1. Replicate the setup described above:
    - LUKS-encrypted root partition, with partition-type GUID allowing for discoverable auto-mounting.
    - Use systemd-cryptenroll to add a FIDO2 hardware token to one of the LUKS partition's keyslots.
    - Do not define an `/etc/crypttab` as these entries will conflict with systemd-gpt-auto-generator
    - Install systemd-boot
    - Copy over the above configs for Dracut
    - Generate a UKI using Dracut
    - Reboot

**Expected behavior**
During boot, when unlocking the LUKS root partition, the user is asked to enter a PIN for the FIDO2 token, after which the user may also confirm physical presence by touching the token (depending on how the token is setup with systemd-cryptenroll).

**Actual behavior**
During boot, the user is instead asked for a passphrase to decrypt the root partition. 

**Additional context**
Cancelling with ctrl+c or ctrl+d reprompts for the passphrase (on previous occasions, I've noticed a race condition where sometimes the YubiKey isn't ready in time, and so the FIDO2 PIN prompt only appears after exiting the passphrase prompt)

**Comment by LaszloGombos** [comment]
@kjkent Can you please try 'rd.auto=0' instead of 'rd.auto=1' ? 

I think the next step would be to add test case to hit the configuration that you have. There were some improvement in the test suite lately. I would suggest to change  https://github.com/dracut-ng/dracut-ng/blob/main/test/TEST-04-FULL-SYSTEMD/test.sh#L40 . This will not only help upstream fix the issue but prevent further regressions.

**Comment by kjkent** [comment]
@LaszloGombos absolutely; I'll give `rd.auto=0` a shot now and unless that fixes things, I'll make a test case. Thanks for the quick response.

**Comment by LaszloGombos** [comment]
@kjkent since you're on Arch, you could also run experiments with https://aur.archlinux.org/packages/dracut-git as the issue might be already resolved (or at least the code around this area might have already changed here upstream). See https://github.com/dracut-ng/dracut-ng/pull/529

**Comment by kjkent** [comment]
@LaszloGombos After some experimenting ⚗️🔬🧪 (also with the `dracut-git` package), I think I've found the issue and, if I'm right, it's nothing to do with Dracut. If so, sorry to have wasted your time.

Looking at the [systemd-cryptsetup docs](https://www.freedesktop.org/software/systemd/man/latest/systemd-cryptsetup.html), under "Credentials":
> ...The following credentials are used by "systemd-crypsetup@root.service" (generated by **systemd-gpt-auto-generator**)...

So, systemd-gpt-auto-generator is responsible for creating systemd-cryptsetup@root.service (something that seems obvious in retrospect; but I was previously unaware of the exact mechanism/sequence of events in the initrd. After rebooting and letting the passphrase prompt fail, I ran `systemctl status systemd-cryptsetup@root.service` and got a similar output to the above.

systemd-gpt-auto-generator's resultant service is ignoring kernel cmdline luks options, and systemd-cryptsetup, when given tpm2-related command line arguments, doesn't check the luks header for fido2 info. In the initrd, if I run `systemd-cryptsetup attach root /dev/gpt-auto-root-luks` manually, I'm immediately asked for the FIDO2 pin, so, everything else is in place.
